File: rubygems/pray-cli/lib/pray/upstream_drift.rb
# frozen_string_literal: true

module Pray
  module Upstream
    module_function

    def path_fork_drift_lines(project, previous, options)
      user_config = Config.load_user_config
      git_sources = GitSources.prepare_git_sources(
        project.project_root,
        project.manifest.sources,
        previous,
        refresh: options.refresh || options.refresh_source_revisions
      )
      sources = Resolve.source_map(project.manifest.sources)
      lines = []
      project.packages.each do |package|
        next unless package.declaration.path
        next unless package.upstream

        upstream = package.upstream
        resolved = resolve_named(
          project.project_root, sources, git_sources, user_config, previous, options,
          upstream.name, "= #{upstream.version}", upstream.source
        )
        upstream_content = content_file_bytes(resolved.root, resolved.spec)
        local_content = local_content_for_refresh(package.root, package.spec)
        if local_content.empty?
          lines << "#{package.declaration.name} has no content files; run pray install to copy #{upstream.name} #{upstream.version}"
          next
        end

        overlay_file_changes(local_content, upstream_content).each do |path, change|
          lines << overlay_drift_line(
            package.declaration.name, upstream.name, upstream.version, path, change
          )
        end
      end
      lines
    end
  end
end

File: npmjs/pray-cli/src/resolve/upstream-drift.ts
import { prepareGitSources } from "../git/sources.js";
import type { Lockfile } from "../lockfile/types.js";
import type { ResolveOptions } from "./context.js";
import { resolvePackage } from "./project.js";
import { sourceMap } from "./source-map.js";
import type { ResolvedProject } from "./types.js";
import { contentFileBytes, localContentForRefresh } from "./upstream-io.js";
import { overlayDriftLine, overlayFileChanges } from "./upstream-merge.js";

export async function pathForkDriftLines(
  project: ResolvedProject,
  previous: Lockfile | undefined,
  options: ResolveOptions,
): Promise<string[]> {
  const gitSources = prepareGitSources(
    project.projectRoot,
    project.manifest.sources,
    previous,
    options.refreshSourceRevisions,
    options.offline,
  );
  const sources = sourceMap(project.manifest.sources);
  const lines: string[] = [];
  for (const packageEntry of project.packages) {
    if (!packageEntry.declaration.path || !packageEntry.upstream) {
      continue;
    }
    const upstream = packageEntry.upstream;
    const declaration = {
      name: upstream.name,
      constraint: `= ${upstream.version}`,
      ...(upstream.source ? { source: upstream.source } : {}),
      exports: [],
      targets: [],
      features: [],
      groups: [],
      optional: false,
      roles: [],
      bound: false,
    };
    const resolved = await resolvePackage(
      project.projectRoot,
      sources,
      gitSources,
      declaration,
      previous,
      options,
    );
    const upstreamContent = contentFileBytes(resolved.root, resolved.spec);
    const localContent = localContentForRefresh(
      packageEntry.root,
      packageEntry.spec,
    );
    if (localContent.size === 0) {
      lines.push(
        `${packageEntry.declaration.name} has no content files; run pray install to copy ${upstream.name} ${upstream.version}`,
      );
      continue;
    }
    for (const [path, change] of overlayFileChanges(
      localContent,
      upstreamContent,
    )) {
      lines.push(
        overlayDriftLine(
          packageEntry.declaration.name,
          upstream.name,
          upstream.version,
          path,
          change,
        ),
      );
    }
  }
  return lines;
}
